File: app/src/test/java/com/codinglikeapirate/pocitaj/ExerciseBookTest.java
package com.codinglikeapirate.pocitaj;

import static org.junit.Assert.*;
import org.junit.Test;

public class ExerciseBookTest {

  @Test
  public void addition_Question() {
    ExerciseBook.Exercise exercise = new ExerciseBook.Addition(2, 3);
    assertEquals("2 + 3", exercise.question());
  }

  @Test
  public void addition_SolveNotRecognized() {
    ExerciseBook.Exercise exercise = new ExerciseBook.Addition(2, 3);
    assertFalse(exercise.solve(ExerciseBook.NOT_RECOGNIZED));
    assertFalse(exercise.solved());
    assertFalse(exercise.correct());
  }

  @Test
  public void addition_SolveIncorrectly() {
    ExerciseBook.Exercise exercise = new ExerciseBook.Addition(2, 3);
    assertFalse(exercise.solve(7));
    assertTrue(exercise.solved());
    assertFalse(exercise.correct());
  }

  @Test
  public void addition_SolveCorrectly() {
    ExerciseBook.Exercise exercise = new ExerciseBook.Addition(2, 3);
    assertTrue(exercise.solve(5));
    assertTrue(exercise.solved());
    assertTrue(exercise.correct());
  }

  @Test
  public void exerciseBook_generatesQuestionAtStart() {
    ExerciseBook exerciseBook = new ExerciseBook();
    assertNotNull(exerciseBook.getLast());
  }

  @Test
  public void exerciseBook_emptyStats() {
    ExerciseBook exerciseBook = new ExerciseBook();
    assertEquals("0 / 0 (0%)", exerciseBook.getStats());
  }

  @Test
  public void exerciseBook_allWrongStats() {
    ExerciseBook exerciseBook = new ExerciseBook();
    int incorrect = 100;
    exerciseBook.getLast().solve(incorrect);
    exerciseBook.generate();
    exerciseBook.getLast().solve(incorrect);
    assertEquals("0 / 2 (0%)", exerciseBook.getStats());
  }

  @Test
  public void exerciseBook_oneSolvedOneUnsolvedWrongStats() {
    ExerciseBook exerciseBook = new ExerciseBook();
    int incorrect = 100;
    exerciseBook.getLast().solve(incorrect);
    exerciseBook.generate();
    assertEquals("0 / 1 (0%)", exerciseBook.getStats());
  }
}

File: app/src/test/java/com/codinglikeapirate/pocitaj/ExerciseBookTest.kt
package com.codinglikeapirate.pocitaj

import com.codinglikeapirate.pocitaj.logic.Addition
import com.codinglikeapirate.pocitaj.logic.Exercise
import com.codinglikeapirate.pocitaj.logic.MissingAddend
import com.codinglikeapirate.pocitaj.logic.MissingSubtrahend
import com.codinglikeapirate.pocitaj.logic.Multiplication
import com.codinglikeapirate.pocitaj.logic.Subtraction
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test


class ExerciseBookTest {

    @Test
    fun addition_Question() {
        val equation = Addition(2, 3)
        assertEquals("2 + 3 = ?", equation.question())
    }

    @Test
    fun subtraction_Question() {
        val equation = Subtraction(5, 3)
        assertEquals("5 - 3 = ?", equation.question())
    }

    @Test
    fun addition_EquationStringBeforeSolving() {
        val equation = Addition(2, 3)
        val exercise = Exercise(equation)
        assertEquals("2 + 3 = ?", exercise.equationString())
    }

    @Test
    fun addition_SolveNotRecognized() {
        val equation = Addition(2, 3)
        assertEquals("2 + 3 = ?", equation.question())

        val exercise = Exercise(equation)
        assertFalse(exercise.solve(Exercise.NOT_RECOGNIZED))
        assertFalse(exercise.solved)
        assertFalse(exercise.correct())
        assertEquals("2 + 3 = ?", exercise.equationString())
    }

    @Test
    fun addition_SolveIncorrectly() {
        val equation = Addition(4, 2)
        assertEquals("4 + 2 = ?", equation.question())

        val exercise = Exercise(equation)
        assertFalse(exercise.solve(7))
        assertTrue(exercise.solved)
        assertFalse(exercise.correct())
        assertEquals("4 + 2 ≠ 7", exercise.equationString())
    }

    @Test
    fun addition_SolveCorrectly() {
        val equation = Addition(2, 3)
        val exercise = Exercise(equation)
        assertTrue(exercise.solve(5))
        assertTrue(exercise.solved)
        assertTrue(exercise.correct())
        assertEquals("2 + 3 = 5", exercise.equationString())
    }


    // Add tests for the equationString() function
    @Test
    fun `equationString returns correct format for unsolved Addition`() {
        val exercise = Exercise(Addition(5, 3))
        assertEquals("5 + 3 = ?", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for solved correct Addition`() {
        val exercise = Exercise(Addition(5, 3))
        exercise.solve(8)
        assertEquals("5 + 3 = 8", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for solved incorrect Addition`() {
        val exercise = Exercise(Addition(5, 3))
        exercise.solve(7)
        assertEquals("5 + 3 ≠ 7", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for unsolved Subtraction`() {
        val exercise = Exercise(Subtraction(10, 4))
        assertEquals("10 - 4 = ?", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for solved correct Subtraction`() {
        val exercise = Exercise(Subtraction(10, 4))
        exercise.solve(6)
        assertEquals("10 - 4 = 6", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for solved incorrect Subtraction`() {
        val exercise = Exercise(Subtraction(10, 4))
        exercise.solve(5)
        assertEquals("10 - 4 ≠ 5", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for unsolved Multiplication`() {
        val exercise = Exercise(Multiplication(7, 3))
        assertEquals("7 × 3 = ?", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for solved correct Multiplication`() {
        val exercise = Exercise(Multiplication(7, 3))
        exercise.solve(21)
        assertEquals("7 × 3 = 21", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for solved incorrect Multiplication`() {
        val exercise = Exercise(Multiplication(7, 3))
        exercise.solve(5)
        assertEquals("7 × 3 ≠ 5", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for unsolved MissingAddend`() {
        val exercise = Exercise(MissingAddend(7, 15))
        assertEquals("7 + ? = 15", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for solved correct MissingAddend`() {
        val exercise = Exercise(MissingAddend(7, 15))
        exercise.solve(8)
        assertEquals("7 + 8 = 15", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for solved incorrect MissingAddend`() {
        val exercise = Exercise(MissingAddend(7, 15))
        exercise.solve(9)
        // For incorrect, it should show the original equation with the incorrect submitted answer
        assertEquals("7 + 9 ≠ 15", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for unsolved MissingSubtrahend`() {
        val exercise = Exercise(MissingSubtrahend(20, 8))
        assertEquals("20 - ? = 8", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for solved correct MissingSubtrahend`() {
        val exercise = Exercise(MissingSubtrahend(20, 8))
        exercise.solve(12)
        // Expected correct format: operand1 - submittedSolution = result
        assertEquals("20 - 12 = 8", exercise.equationString())
    }

    @Test
    fun `equationString returns correct format for solved incorrect MissingSubtrahend`() {
        val exercise = Exercise(MissingSubtrahend(20, 8))
        exercise.solve(10)
        // For incorrect, it should show the original equation with the incorrect submitted answer
        assertEquals("20 - 10 ≠ 8", exercise.equationString())
    }
}